### DISCUSSION TITLE
Thinking about the future of releases...

### DISCUSSION THREAD

**Discussion by Potherca** [plantuml-stdlib/C4-PlantUML#293] [General]
Over the last few years, development of C4-PlantUML has become mostly stable (largely through the efforts of @kirchsth).

Looking back, the release schedule was this:

| Version                                                                                                              | Release Date |
| ----------------------------------------------------------------------------------------------------- | ------------------ |
| [v1.0.0](https://github.com/plantuml-stdlib/C4-PlantUML/releases/tag/v1.0.0) |   2018-12-14  |
| [v2.0.0](https://github.com/plantuml-stdlib/C4-PlantUML/releases/tag/v2.0.0) |   2020-11-28  |
| [v2.0.1](https://github.com/plantuml-stdlib/C4-PlantUML/releases/tag/v2.0.1) |   2021-02-08  |
| [v2.1.0](https://github.com/plantuml-stdlib/C4-PlantUML/releases/tag/v2.1.0) |   2021-04-03  | 
| [v2.2.0](https://github.com/plantuml-stdlib/C4-PlantUML/releases/tag/v2.2.0) |   2021-04-11  | 
| [v2.3.0](https://github.com/plantuml-stdlib/C4-PlantUML/releases/tag/v2.3.0) |   2021-07-04  |
| [v2.4.0](https://github.com/plantuml-stdlib/C4-PlantUML/releases/tag/v2.4.0) |   2021-10-29  | 
| [v2.5.0](https://github.com/plantuml-stdlib/C4-PlantUML/releases/tag/v2.5.0) |   2022-12-11  |

Looking forward, we might want to formulate some kind of rules or something of a plan as to what we would like to release when.

Not only for larger features and fixes, but also to get small changes (like the recent #292) into the upstream PlantUML-stdlib faster.

<sup>(Part of this is also #148 and #276)</sup>

Anyone else have thoughts or opinions on this topic?

**Comment by kirchsth** [comment]
1. we could create for minor changes with an improved functionality a patch version:
  e.g. #281 (sprites with empty label and better link support).  
  I think #292 is too small (it's only a comment typo in C4.puml, all other are not part of the  PlantUML-stdlib)

2. what means patch?
   That we can have a new PlantUML-stdlib version we need at least a new branch that we can update the `C4Version()`.
   and start with a `release-notes.md(?)` that we can document it.  
   Practically I think we will end with the same steps/effort as for a normal release.

3. As long we have no automatic process I think we could make smaller/time-based milestones
   - E.g. all 3/6 month a new version which is not assigned to a concrete version,
     and at the end of the 3/6 month create a patch or (minor) version depending of the content.
   - All planned (and assigned) issues which are not planed in this 3/6 month are moved to a "Backlog"
     milestone and labeled with `not-stale`
  - And if we have multiple bigger improvements before we reach the 3/6 month we could make an
     additional intermediate version too 
  - Practically we rename the v2.6.0 milestone to "2023-Q2", move all finished or realistic finished (until end of Q2)
    in this "2023-Q2" milestone and all other into the "backlog" milestone.  
    And depending on the finished content we create a v2.5.1 or v2.6.0 version.
    If we are really fast we could  have a v2.6.0 in May and v2.6.1 end of June.

**Reply by kirchsth** [reply]
Hi @Potherca,
I think the pattern with the 2 milestones is too complex. At the end we will have a new release if we have enough relevant changes (or maybe if the gap between the releases is too big).

I would make a new 2.7.0 release with the sequence diagrams, the reference to mweagle themes and PlantUML has the improved !theme support (!theme C4_united from <C4/themes>) in the meantime too.

But after that I would define/use only one dummy milestone that we can collect issues and MRs.

What do you think?
BR Helmut

**Reply by Potherca** [reply]
> I think the pattern with the 2 milestones is too complex. 

I am inclined to agree. The pattern originally comes from my work in other repos, but it is based on the assumption that there is someone who makes decision regarding prioritites, which is not the process here.
